Implement font-variant-alternates matching

mdubet opened the pull request:
#### a14a556da51f33eb6557c948e684c7f783d5da77
<pre>
Implement font-variant-alternates matching
<a href="https://bugs.webkit.org/show_bug.cgi?id=246121">https://bugs.webkit.org/show_bug.cgi?id=246121</a>
rdar://100830615

Reviewed by Myles C. Maxfield.

There are still some parts of the spec missing:
  * mutiple ident in font-variant-alternates for styleset and character-variant
  * serialization of overwritten declarations
The FontFeatureValues object is reference-counted
because it is fairly large (it contains 6 HashMaps).

* LayoutTests/TestExpectations:
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::font):
* Source/WebCore/css/CSSFontFace.h:
* Source/WebCore/css/CSSFontFeatureValuesRule.cpp:
(WebCore::CSSFontFeatureValuesRule::cssText const):
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::addFontPaletteValuesRule):
(WebCore::CSSFontSelector::addFontFeatureValuesRule):
(WebCore::CSSFontSelector::lookupFontFeatureValues):
(WebCore::CSSFontSelector::fontRangesForFamily):
* Source/WebCore/css/CSSFontSelector.h:
* Source/WebCore/css/CSSSegmentedFontFace.cpp:
(WebCore::CSSSegmentedFontFace::fontRanges):
* Source/WebCore/css/CSSSegmentedFontFace.h:
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleFontFeatureValues::StyleRuleFontFeatureValues):
(WebCore::StyleRuleFontFeatureValuesBlock::StyleRuleFontFeatureValuesBlock):
(WebCore::StyleRuleFontFeatureValues::create):
* Source/WebCore/css/StyleRule.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::fontFeatureValuesTypeMappings):
(WebCore::CSSParserImpl::consumeFontFeatureValuesRuleBlock):
(WebCore::CSSParserImpl::consumeFontFeatureValuesRule):
* Source/WebCore/platform/graphics/FontCreationContext.h:
(WebCore::FontCreationContextRareData::create):
(WebCore::FontCreationContextRareData::fontFeatureValues const):
(WebCore::FontCreationContextRareData::operator== const):
(WebCore::FontCreationContextRareData::FontCreationContextRareData):
(WebCore::FontCreationContext::FontCreationContext):
(WebCore::FontCreationContext::fontFeatureValues const):
(WebCore::add):
* Source/WebCore/platform/graphics/FontFeatureValues.cpp: Added.
(WebCore::add):
(WebCore::FontFeatureValues::updateOrInsert):
(WebCore::FontFeatureValues::updateOrInsertForType):
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/FontFeatureValues.h: Added.
(WebCore::FontFeatureValues::create):
(WebCore::FontFeatureValues::isEmpty const):
(WebCore::FontFeatureValues::operator== const):
(WebCore::FontFeatureValues::operator!= const):
(WebCore::FontFeatureValues::styleset const):
(WebCore::FontFeatureValues::styleset):
(WebCore::FontFeatureValues::stylistic const):
(WebCore::FontFeatureValues::stylistic):
(WebCore::FontFeatureValues::characterVariant const):
(WebCore::FontFeatureValues::characterVariant):
(WebCore::FontFeatureValues::swash const):
(WebCore::FontFeatureValues::swash):
(WebCore::FontFeatureValues::ornaments const):
(WebCore::FontFeatureValues::ornaments):
(WebCore::FontFeatureValues::annotation const):
(WebCore::FontFeatureValues::annotation):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::preparePlatformFont):
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
* Source/WebCore/platform/graphics/harfbuzz/ComplexTextControllerHarfBuzz.cpp:
(WebCore::fontFeatures):
* Source/WebCore/platform/text/TextFlags.cpp:
(WebCore::operator&lt;&lt;):
(WebCore::add):
(WebCore::computeFeatureSettingsFromVariants):
(WebCore::reduce): Deleted.
* Source/WebCore/platform/text/TextFlags.h:
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRules):
(WebCore::Style::RuleSetBuilder::addMutatingRulesToResolver):

Canonical link: <a href="https://commits.webkit.org/256002@main">https://commits.webkit.org/256002@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0e33ec7ba828420cfa7b9ce8982c22b17374cd2d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94167 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3356 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24692 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103789 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164120 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98157 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3378 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31572 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86474 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99808 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99825 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2426 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80579 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29448 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84351 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/84030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72396 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37956 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17852 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35841 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19124 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4152 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39714 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41708 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41658 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38355 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/platform/graphics/FontFeatureValues.cpp:
```diff
@@ -0,0 +1,84 @@
+    upsertTags(characterVariant, toUpsert->characterVariant);
+    upsertTags(swash, toUpsert->swash);
+    upsertTags(ornaments, toUpsert->ornaments);
+    upsertTags(annotation, toUpsert->annotation);
```
I've never heard of `upsert` until now, it's in fact never used in the WebKit codebase aside from the code you've added. I think the usual terminology here is just `set`.

mdubet replied:
It's a fairly common term: https://wiki.postgresql.org/wiki/UPSERT https://en.wiktionary.org/wiki/upsert
It conveys more meaning than "set"

nt1m replied:
Seems common in the database world, not really in WebKit.
In fact, our own HashMap in WTF uses `set` as terminology

nt1m replied:
(Even JS Map() uses `set`)

mdubet replied:
But it's not a `set`. 

If I have `this` object with tags `[foo=>1; bar=>2]` and the `other` object with tags `[foo=>5, lol=>10]`.

`this.set(B)` would result in `this` being exactly `[foo=>5; lol=>10]`.
`this.upsert(B)` would result in `this` being `[foo=>5; bar=>2; lol=>10]`.

nt1m replied:
`updateOrInsert()` seems more explicit then.

mdubet replied:
Well this is exactly how the term `upsert` is constructed indeed (UPdate_or_inSERT).

nt1m replied:
My concern is mostly the lack of usage in WebKit: https://searchfox.org/wubkat/search?q=upsert&path=&case=false&regexp=false

whsieh replied:
As Tim noted, I think `updateOrInsert` is likely the preferred name in WebKit, per style guidelines that prefer [full words over abbreviations](https://webkit.org/code-style-guidelines/#names) or (in this case) portmanteaus. I'm not sure this counts as one of the cases where the abbreviation is easier to understand, given that there's no other usage in WebKit, and it doesn't reflect the name of a DOM API (like `tabIndex` in the style guidelines example).

Perhaps, if we want to keep a short name, we could consider `replace`, or `merge`?

mdubet replied:
I might be biased by my background in Postgres world 🙃 I will follow your opinions and change it to `updateOrInsert`.

nt1m commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -718,27 +718,28 @@ RefPtr<StyleRuleFontFeatureValues> CSSParserImpl::consumeFontFeatureValuesRule(C
+    // Convert block rules to value (remove duplicate...etc)
+    auto fontFeatureValues = FontFeatureValues::create();
+
+    auto upsertTag = [&fontFeatureValues](FontFeatureValuesType type, Tag tag) {
```
I know it was there before, but can we also rename this?

mdubet replied:
Fixed

nt1m commented on LayoutTests/platform/gtk/TestExpectations:
```diff
@@ -1960,4 +1960,20 @@ webkit.org/b/214463 imported/w3c/web-platform-tests/css/css-sizing/contain-intri
+webkit.org/b/246121 imported/w3c/web-platform-tests/css/css-fonts/font-variant-alternates-15.html [ ImageOnlyFailure ]
+webkit.org/b/246121 imported/w3c/web-platform-tests/css/css-fonts/font-variant-alternates-16.html [ ImageOnlyFailure ]
+webkit.org/b/246121 imported/w3c/web-platform-tests/css/css-fonts/font-variant-alternates-17.html [ ImageOnlyFailure ]
+webkit.org/b/246121 imported/w3c/web-platform-tests/css/css-fonts/font-variant-alternates-18.html [ ImageOnlyFailure ]
```
Please file a bug for GTK and reference that bug instead.

mdubet replied:
Fixed https://bugs.webkit.org/show_bug.cgi?id=246952

litherum commented on Source/WebCore/css/CSSFontFace.cpp:
```diff
@@ -710,7 +710,7 @@ static Font::Visibility visibility(CSSFontFace::Status status, CSSFontFace::Font
 }
 
-RefPtr<Font> CSSFontFace::font(const FontDescription& fontDescription, bool syntheticBold, bool syntheticItalic, ExternalResourceDownloadPolicy policy, const FontPaletteValues& fontPaletteValues)
+RefPtr<Font> CSSFontFace::font(const FontDescription& fontDescription, bool syntheticBold, bool syntheticItalic, ExternalResourceDownloadPolicy policy, const FontPaletteValues& fontPaletteValues, const Ref<FontFeatureValues> fontFeatureValues)
```
Does it make sense to put palette-values and feature-values together in a struct and thread that struct around, rather than having many arguments passed around?

darinadler replied:
I think that’s a good suggestion.

Separately, the argument type `const Ref<FontFeatureValues>` does not make sense. You could take const Ref<FontFeatureValues>&, or `Ref<FontFeatureValues>`, or `Ref<FontFeatureValues>&&`, or `FontFeatureValues&`. But the `const` doesn’t do anything useful. It’s like having an argument of type `const bool`.

mdubet replied:
I guess it would. There is already the `FontCreationContext` class which kinda serves the same point but not in this part of the codebase.

litherum commented on Source/WebCore/css/CSSFontFeatureValuesRule.cpp:
```diff
@@ -51,7 +51,7 @@ String CSSFontFeatureValuesRule::cssText() const
     joinFontFamiliesWithSeparator(m_fontFeatureValuesRule->fontFamilies(), ", "_s);
     builder.append(" { ");
-    auto& value = m_fontFeatureValuesRule->value();
+    auto value = m_fontFeatureValuesRule->value();
```
Is this necessary?

mdubet replied:
No indeed, I've fixed it to `const&`

litherum commented on Source/WebCore/platform/graphics/FontFeatureValues.h:
```diff
@@ -0,0 +1,91 @@
+    Tags characterVariant;
+    Tags swash;
+    Tags ornaments;
+    Tags annotation;
```
If members are public, it should be a struct, rather than a class.

litherum commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -477,15 +477,15 @@ RetainPtr<CTFontRef> preparePlatformFont(CTFontRef originalFont, const FontDescr
 
     // Step 7: Consult with font-feature-settings inside @font-face
-    if (fontCreationContext.fontFaceFeatures() && !fontCreationContext.fontFaceFeatures()->isEmpty()) {
+    if (fontCreationContext.fontFaceFeatures()) {
```
Why?

mdubet replied:
This condition is already handled on the next line with the `for` loop. It seems like a bit of duplication.

litherum commented on Source/WebCore/platform/graphics/harfbuzz/ComplexTextControllerHarfBuzz.cpp:
```diff
@@ -195,9 +195,9 @@ static Vector<hb_feature_t, 4> fontFeatures(const FontCascade& font, const FontP
     // 3. Font features implied by the value of the ‘font-variant’ property, the related ‘font-variant’
     //    subproperties and any other CSS property that uses OpenType features.
-    for (auto& feature : computeFeatureSettingsFromVariants(font.fontDescription().variantSettings()))
+    for (auto& feature : computeFeatureSettingsFromVariants(font.fontDescription().variantSettings(), { }))
```
Might want to add a FIXME or something, with a bug URL

mdubet replied:
Added

litherum commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -278,13 +264,60 @@ FeaturesMap computeFeatureSettingsFromVariants(const FontVariantSettings& varian
+            addFeatureTagWithValue(values.swash, fontFeatureValues->swash, "cswh");
+
+            addFeatureTagWithValue(values.ornaments, fontFeatureValues->ornaments, "ornm");
+            addFeatureTagWithValue(values.annotation, fontFeatureValues->annotation, "nalt");
```
🆒

litherum reviewed (approved):
This is a pretty big patch, and isn't super focused. In the future, I'd recommend making more, and smaller, patches. It helps reviewability and helps bisections in the future.

darinadler commented on Source/WebCore/css/CSSFontSelector.cpp:
```diff
@@ -245,6 +246,23 @@ void CSSFontSelector::addFontPaletteValuesRule(StyleRuleFontPaletteValues& fontP
+{
+    Ref<FontFeatureValues> fontFeatureValues = fontFeatureValuesRule.value();
+
+    for (auto fontFamily : fontFeatureValuesRule.fontFamilies()) {
```
`auto&` is better here

darinadler commented on Source/WebCore/css/CSSFontSelector.cpp:
```diff
@@ -245,6 +246,23 @@ void CSSFontSelector::addFontPaletteValuesRule(StyleRuleFontPaletteValues& fontP
+            initial = exist->value;
+
+        initial->updateOrInsert(fontFeatureValues);
+        m_featureValues.set(fontFamily, initial);
```
This is the wrong idiom for updating in place, inefficient and kind of wordy too. Instead we should use `add` or `ensure` and then call `updateOrInsert(fontFeatureValues)` on the value in place inside the hash table.

darinadler commented on Source/WebCore/css/CSSFontSelector.h:
```diff
@@ -102,6 +104,7 @@ class CSSFontSelector final : public FontSelector, public CSSFontFace::Client, p
     std::optional<AtomString> resolveGenericFamily(const FontDescription&, const AtomString& family);
 
     const FontPaletteValues& lookupFontPaletteValues(const AtomString& familyName, const FontDescription&);
+    Ref<FontFeatureValues> lookupFontFeatureValues(const AtomString& familyName);
```
The word "lookup" is a noun, but this and the function above use it as a verb. That should be two words, "look up", so written `lookUp` instead. I also don’t understand why we need to call it "look up"; usually we just call this "find" or don’t use a verb at all.

mdubet replied:
I agree, but I think that there is some value in keeping the wording already shipped. Having both `lookup` and `lookUp` functions right next to each other feels kinda wrong. Maybe we should change them both in a followup patch ?

darinadler commented on Source/WebCore/css/CSSFontSelector.h:
```diff
@@ -67,7 +68,8 @@ class CSSFontSelector final : public FontSelector, public CSSFontFace::Client, p
 
     void addFontFaceRule(StyleRuleFontFace&, bool isInitiatingElementInUserAgentShadowTree);
-    void addFontPaletteValuesRule(StyleRuleFontPaletteValues&);
+    void addFontPaletteValuesRule(const StyleRuleFontPaletteValues&);
+    void addFontFeatureValuesRule(const StyleRuleFontFeatureValues&);
```
It’s not helpful to include the type names in the names of all these functions. They should just be named `add` or `addRule`. C++ will select the correct function based on the argument type.

mdubet replied:
Same, I'm following the already shipped naming pattern. A renaming only follow-up patch could be a solution though.

darinadler commented on Source/WebCore/css/CSSFontSelector.cpp:
```diff
@@ -351,6 +380,7 @@ FontRanges CSSFontSelector::fontRangesForFamily(const FontDescription& fontDescr
     };
 
     const auto& fontPaletteValues = lookupFontPaletteValues(familyName, fontDescription);
+    Ref<FontFeatureValues> fontFeatureValues = lookupFontFeatureValues(familyName);
```
It’s really unfortunate that we have designed an interface that always allocates these on the heap, and requires allocating a destroying an empty one just to express "there are no values". We could use `RefPtr` instead and use nullptr for when there are none, or maybe not heap-allocate `FontFeatureValues`?

mdubet replied:
`FontFeatureValues` is big and thus kinda need to be heap-allocated (for example, there is a `static_assert` related to key of HashMap being smaller than its size).

Indeed, we need to express the fairly frequent "there is no value" : I've changed the patch to use `RefPtr` instead of `Ref` in most places. We would not have to allocate/deallocate the empty one anymore.

darinadler commented on Source/WebCore/css/CSSSegmentedFontFace.cpp:
```diff
@@ -60,26 +60,27 @@ void CSSSegmentedFontFace::fontLoaded(CSSFontFace&)
 class CSSFontAccessor final : public FontAccessor {
 public:
-    static Ref<CSSFontAccessor> create(CSSFontFace& fontFace, const FontDescription& fontDescription, const FontPaletteValues& fontPaletteValues, bool syntheticBold, bool syntheticItalic)
+    static Ref<CSSFontAccessor> create(CSSFontFace& fontFace, const FontDescription& fontDescription, const FontPaletteValues& fontPaletteValues, const Ref<FontFeatureValues> fontFeatureValues, bool syntheticBold, bool syntheticItalic)
```
Same `const Ref<>` mistake I mentioned above.

darinadler commented on Source/WebCore/css/CSSSegmentedFontFace.h:
```diff
@@ -50,7 +50,7 @@ class CSSSegmentedFontFace final : public RefCounted<CSSSegmentedFontFace>, publ
     void appendFontFace(Ref<CSSFontFace>&&);
 
-    FontRanges fontRanges(const FontDescription&, const FontPaletteValues&);
+    FontRanges fontRanges(const FontDescription&, const FontPaletteValues&, const Ref<FontFeatureValues>);
```
Here too.

darinadler commented on Source/WebCore/css/StyleRule.h:
```diff
@@ -227,22 +207,22 @@ class StyleRuleFontFeatureValuesBlock final : public StyleRuleBase {
     const Vector<AtomString>& fontFamilies() const { return m_fontFamilies; }
 
-    const FontFeatureValuesValue& value() const { return m_value; }
+    const Ref<FontFeatureValues> value() const { return m_value; }
```
This doesn’t make sense as a return type. We could return `Ref<>` or `const Ref<>&`, but not `const Ref<>`.

darinadler commented on Source/WebCore/platform/graphics/FontCreationContext.h:
```diff
@@ -49,10 +50,16 @@ class FontCreationContextRareData : public RefCounted<FontCreationContextRareDat
         return m_fontPaletteValues;
     }
 
+    const Ref<FontFeatureValues> fontFeatureValues() const
```
Same problem here with `const Ref<>`.

darinadler commented on Source/WebCore/platform/graphics/FontCreationContext.h:
```diff
@@ -98,6 +106,11 @@ class FontCreationContext {
         return m_rareData ? &m_rareData->fontPaletteValues() : nullptr;
     }
 
+    const RefPtr<FontFeatureValues> fontFeatureValues() const
```
Same mistake here with `const RefPtr`.

darinadler commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -61,48 +61,36 @@ WTF::TextStream& operator<<(TextStream& ts, Kerning kerning)
+            if (!builder.isEmpty())
+                builder.append(" ");
+
+            builder.append(std::forward<Ts>(args)...);
```
Better efficiency:
```
    builder.append(builder.isEmpty() ? "" : " ", std::forward<Ts>(args)...);
```

darinadler commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -278,13 +264,60 @@ FeaturesMap computeFeatureSettingsFromVariants(const FontVariantSettings& varian
         break;
     }
 
+    // We don't do anything if "normal".
```
This is not a helpful comment. It just says exactly what the line below says. A helpful comment might answer the question "why?" or we could omit the comment.

darinadler commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -278,13 +264,60 @@ FeaturesMap computeFeatureSettingsFromVariants(const FontVariantSettings& varian
+
+            // Some features (styleset, character-variant and historical-forms) don't have values,
+            // the tag name itself is the actual conveyor of information.
+            auto addFeatureTags = [&features, &lookupTags](const auto& name, const auto& tags, auto codename) {
```
I think the type of the codename should not just be "auto". We want a compile time error if we pass something that is less than 2 digits long.

mdubet replied:
Fixed to `std::array<char ,2>`

darinadler commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -278,13 +264,60 @@ FeaturesMap computeFeatureSettingsFromVariants(const FontVariantSettings& varian
+                    char rightDigit = '0' + (value % 10);
+                    char leftDigit = '0' + (value / 10);
+                    std::string s = { codename[0], codename[1], leftDigit, rightDigit };
+                    features.set(fontFeatureTag(s.c_str()), 1);
```
We should not use `std::string` here, which allocates on the heap. This can just be a `std::array` on the stack since it’s a fixed length.

darinadler commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -278,13 +264,60 @@ FeaturesMap computeFeatureSettingsFromVariants(const FontVariantSettings& varian
+                if (found == tags.end())
+                    return { };
+
+                return found->value;
```
This is what `get()` does, no need to use `find()`/`end()`.

darinadler commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -278,13 +264,60 @@ FeaturesMap computeFeatureSettingsFromVariants(const FontVariantSettings& varian
+            };
+
+            auto addFeatureTagWithValue = [&features, &lookupTags] (const auto& name, const auto& tags, auto codename) {
+                for (const auto& value : lookupTags(name, tags)) {
```
I suggest we use `int` here instead of `const auto&`.

darinadler commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -278,13 +264,60 @@ FeaturesMap computeFeatureSettingsFromVariants(const FontVariantSettings& varian
+            auto addFeatureTagWithValue = [&features, &lookupTags] (const auto& name, const auto& tags, auto codename) {
+                for (const auto& value : lookupTags(name, tags)) {
+                    // The spec says that we ignore value greater than 99.
+                    if (value > 99)
```
It’s unclear what prevents this from being `< 0`.

mdubet replied:
Indeed, it should be `unsigned` from the beginning (from the parsing). I've fixed it.

darinadler commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -278,13 +264,60 @@ FeaturesMap computeFeatureSettingsFromVariants(const FontVariantSettings& varian
-        // TODO: handle other tags.
-        // https://bugs.webkit.org/show_bug.cgi?id=246121
+        if (fontFeatureValues) {
+            auto lookupTags = [](const auto& name, const auto& tags) -> Vector<int> {
```
It’s really inefficient that we have to copy a vector to return the vector that is sitting in the hash table. We can instead use `Span<const int>` for the return value and then we won’t be allocating and destroying temporary vectors all the time.

darinadler commented on Source/WebCore/platform/text/TextFlags.h:
```diff
@@ -192,6 +193,8 @@ struct FontVariantAlternatesValues {
     }
 
     std::optional<String> stylistic;
+    // TODO: supports a list of string for styleset and characterVariant.
```
WebKit uses `FIXME`, not `TODO`.

darinadler commented on Source/WebCore/platform/text/TextFlags.h:
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include "FontFeatureValues.h"
```
Don’t need to include the whole header just to define an argument of `RefPtr<FontFeatureValues>`. Use a forward declaration instead please.

darinadler commented on Source/WebCore/platform/text/TextFlags.h:
```diff
@@ -483,6 +486,6 @@ enum class AllowUserInstalledFonts : uint8_t {
 
 using FeaturesMap = HashMap<FontTag, int, FourCharacterTagHash, FourCharacterTagHashTraits>;
-FeaturesMap computeFeatureSettingsFromVariants(const FontVariantSettings&);
+FeaturesMap computeFeatureSettingsFromVariants(const FontVariantSettings&, const RefPtr<FontFeatureValues>);
```
The `const` here makes no sense.

mdubet committed:
Implement font-variant-alternates matching
https://bugs.webkit.org/show_bug.cgi?id=246121
rdar://100830615

Reviewed by Myles C. Maxfield.

There are still some parts of the spec missing:
  * mutiple ident in font-variant-alternates for styleset and character-variant
  * serialization of overwritten declarations
The FontFeatureValues object is reference-counted
because it is fairly large (it contains 6 HashMaps).

* LayoutTests/TestExpectations:
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::font):
* Source/WebCore/css/CSSFontFace.h:
* Source/WebCore/css/CSSFontFeatureValuesRule.cpp:
(WebCore::CSSFontFeatureValuesRule::cssText const):
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::addFontPaletteValuesRule):
(WebCore::CSSFontSelector::addFontFeatureValuesRule):
(WebCore::CSSFontSelector::lookupFontFeatureValues):
(WebCore::CSSFontSelector::fontRangesForFamily):
* Source/WebCore/css/CSSFontSelector.h:
* Source/WebCore/css/CSSSegmentedFontFace.cpp:
(WebCore::CSSSegmentedFontFace::fontRanges):
* Source/WebCore/css/CSSSegmentedFontFace.h:
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleFontFeatureValues::StyleRuleFontFeatureValues):
(WebCore::StyleRuleFontFeatureValuesBlock::StyleRuleFontFeatureValuesBlock):
(WebCore::StyleRuleFontFeatureValues::create):
* Source/WebCore/css/StyleRule.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::fontFeatureValuesTypeMappings):
(WebCore::CSSParserImpl::consumeFontFeatureValuesRuleBlock):
(WebCore::CSSParserImpl::consumeFontFeatureValuesRule):
* Source/WebCore/platform/graphics/FontCreationContext.h:
(WebCore::FontCreationContextRareData::create):
(WebCore::FontCreationContextRareData::fontFeatureValues const):
(WebCore::FontCreationContextRareData::operator== const):
(WebCore::FontCreationContextRareData::FontCreationContextRareData):
(WebCore::FontCreationContext::FontCreationContext):
(WebCore::FontCreationContext::fontFeatureValues const):
(WebCore::add):
* Source/WebCore/platform/graphics/FontFeatureValues.cpp: Added.
(WebCore::add):
(WebCore::FontFeatureValues::updateOrInsert):
(WebCore::FontFeatureValues::updateOrInsertForType):
(WebCore::operator<<):
* Source/WebCore/platform/graphics/FontFeatureValues.h: Added.
(WebCore::FontFeatureValues::create):
(WebCore::FontFeatureValues::isEmpty const):
(WebCore::FontFeatureValues::operator== const):
(WebCore::FontFeatureValues::operator!= const):
(WebCore::FontFeatureValues::styleset const):
(WebCore::FontFeatureValues::styleset):
(WebCore::FontFeatureValues::stylistic const):
(WebCore::FontFeatureValues::stylistic):
(WebCore::FontFeatureValues::characterVariant const):
(WebCore::FontFeatureValues::characterVariant):
(WebCore::FontFeatureValues::swash const):
(WebCore::FontFeatureValues::swash):
(WebCore::FontFeatureValues::ornaments const):
(WebCore::FontFeatureValues::ornaments):
(WebCore::FontFeatureValues::annotation const):
(WebCore::FontFeatureValues::annotation):
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::preparePlatformFont):
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
* Source/WebCore/platform/graphics/harfbuzz/ComplexTextControllerHarfBuzz.cpp:
(WebCore::fontFeatures):
* Source/WebCore/platform/text/TextFlags.cpp:
(WebCore::operator<<):
(WebCore::add):
(WebCore::computeFeatureSettingsFromVariants):
(WebCore::reduce): Deleted.
* Source/WebCore/platform/text/TextFlags.h:
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRules):
(WebCore::Style::RuleSetBuilder::addMutatingRulesToResolver):

Canonical link: https://commits.webkit.org/256002@main
